upgrades/man/ose-upgrade.8
.\" Text automatically generated by txt2man
.TH ose-upgrade  "26 November 2013" "" ""
.SH NAME
\fB
\fBose-upgrade
\fB
.SH SYNOPSIS
.nf
.fam C

\fBose-upgrade\fP [status|all|<step> [<step>\.\.\.]]
  [\fB-s\fP|\fB--skip\fP] [\fB-q\fP|\fB--quiet\fP] [\fB-d\fP|\fB--debug\fP] [\fB--complete\fP \fIINTEGER\fP]
  [\fB-h\fP|\fB--help\fP]


.fam T
.fi
.fam T
.fi
.SH DESCRIPTION

\fBose-upgrade\fP is a tool for upgrading an OpenShift Enterprise host
when that upgrade requires coordination and/or changes more substantial
than a simple yum update. 
.SH OPTIONS

.TP
.B
\fB-h\fP|\fB--help\fP
Show a usage summary and exit.
.TP
.B
\fB-q\fP|\fB--quiet\fP
Only display warnings and failures. The default output can be rather verbose.
All output is still logged to /var/log/openshift/upgrade.log even with this option.
.TP
.B
\fB-s\fP|\fB--skip\fP
Mark upgrade \fBstep\fP(s) complete rather than running. Used mainly for
testing or when a problem with a step has been resolved manually.
.TP
.B
\fB-d\fP|\fB--debug\fP
Enable extra debugging output.
.TP
.B
\fB--complete\fP \fIINTEGER\fP
Mark given upgrade number as complete. This clears all upgrade state
in preparation for the next upgrade. Upgrades are numbered beginning with 1,
which was the upgrade of OpenShift Enterprise 1.1 to 1.2.
.SH USAGE

An upgrade consists of one or more steps, each of which often consists of one or more
scripts, all of which must run sequentially. The state of the last completed or current
upgrade is stored in the state file /etc/openshift/upgrade/state.yaml, indicating completed
steps and scripts for the current upgrade. The upgrade log is stored in
/var/log/openshift/upgrade.log.
.PP
Consult the OpenShift Enterprise release notes for step details for specific releases.
.IP \(bu 3
With no arguments, \fBose-upgrade\fP displays status as if "status" had been requested.
.IP \(bu 3
With "status", \fBose-upgrade\fP displays the current upgrade number, the known steps, and the next step.
If no upgrade is pending or in progress, the previous upgrade number is displayed.
.IP \(bu 3
With "all", \fBose-upgrade\fP will run all available steps for this host. Note that part of the upgrade
is likely to require a manual update of \fBose-upgrade\fP itself, after which more steps may become available.
If any step fails, \fBose-upgrade\fP exits at that time.
.IP \(bu 3
Given one or more steps, \fBose-upgrade\fP will run the given steps in order. A step must succeed or be
skipped before the next step can be run.
.PP
Normally \fBose-upgrade\fP exits with exit code 0 for success. A non-zero exit code indicates
a failure to complete a step.
.SH FAILURE

\fBose-upgrade\fP is designed to enable retrying in the event a failure is detected. Nearly all
of the upgrade logic is implemented with scripts that perform a specific update and check
for success; usually these can safely run again until they succeed. Examine the upgrade.log to
determine what failed and how to resolve it (for
example, if a filesystem runs out of space, clear some space or expand the filesystem).
Then run \fBose-upgrade\fP with the failed step again; it will continue running the step/script that
failed, without re-running any that succeeded.
.SH WARNINGS

Prior to an upgrade, it is a good idea to execute oo-admin-chk on a broker and oo-diagnostics
on every deployed host. Address any existing problems prior to using \fBose-upgrade\fP.
.PP
An upgrade requires an outage and generally makes irreversible changes. Read the release notes
to understand what an upgrade entails, and practice upgrading a test deployment if possible.
Ensure that your entire deployment is adequately backed up and can be recovered in the event
of an unexpected failure during an upgrade.
.SH FILES

/var/log/openshift/upgrade.log - a complete log of the host upgrade output for troubleshooting purposes.
.PP
/etc/openshift/upgrade/state.yaml - a YAML file that records the status of the current upgrade,
including steps and scripts that completed or failed.
.SH SEE ALSO

\fBoo-admin-chk\fP(8), \fBoo-accept-broker\fP(8), \fBoo-accept-systems\fP(8),
\fBoo-accept-node\fP(8), \fBoo-diagnostics\fP(8)
.SH AUTHOR
Luke Meyer <lmeyer@redhat.com> - man page written for OpenShift


upgrades/lib/ose-upgrade.rb

#--
# Copyright 2013 Red Hat, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#++

module OSEUpgrader
  OUT_OF_SEQUENCE = 255
  class Abstract

    def initialize(params = {})
      @params = params
    end

    # steps this upgrader can perform, in the order needed
    def implemented_steps
      []
    end

    # what we are trying to upgrade - generic host, broker, node
    def self.upgrade_target; "host"; end
    def upgrade_target; self.class.upgrade_target; end

    def initial_state(upgrade_number = nil)
      { }
    end

    def upgrade_state=(state)
      @upgrade_state = state
    end

    def set_step_status(step, to)
      state = get_step_state_data(step)
      state['previous_status'] = state['status']
      verbose "Setting #{self.upgrade_target} step '#{step}' status to #{to}"
      state['status'] = to
    end

    def get_step_state_data(step)
      steps_hash = @upgrade_state['steps'] ||= {}
      target_hash = steps_hash[upgrade_target()] ||= {}
      return target_hash[step] ||= {'status' => 'NONE'}
    end

    def get_step_status(step)
      return get_step_state_data(step)['status']
    end

    def get_next_step
      step = nil
      self.implemented_steps.each do |s|
        case get_step_status(s)
        when 'NONE', 'FAILED'
          step = s
          break
        when 'UPGRADING'
          do_fail "SHOULD NOT HAPPEN: step '#{s}' has state 'UPGRADING'"
          exit 1
        end
      end
      return step # nil means "nothing left to do here"
    end

    def steps_complete?
      get_next_step.nil?
    end

    def run_upgrade_step(step)
      rc =
      if implemented_steps.include? step
        next_step = self.get_next_step
        if next_step.nil?
          do_warn "All steps complete - no need to re-run step '#{step}'"
          OUT_OF_SEQUENCE
        elsif get_step_status(step) == 'COMPLETE'
          do_warn "Already ran step '#{step}' - please run step '#{next_step}' next"
          OUT_OF_SEQUENCE
        elsif next_step != step
          do_warn "Not ready for step '#{step}' yet - please run step '#{next_step}'"
          OUT_OF_SEQUENCE
        elsif @params[:skip]
          set_step_status(step, 'COMPLETE')
          if step = get_next_step
            verbose "Next step is '#{step}'"
          end
          0
        else
          begin
            set_step_status(step, 'UPGRADING')
            rc = self.send("run_upgrade_step_#{step}", @upgrade_state)
            set_step_status(step, (rc==0)?'COMPLETE':'FAILED')
            if rc==0 and step = get_next_step
              verbose "Next step is '#{step}'"
            end
            rc
          rescue Exception => e
            set_step_status(step, 'FAILED')
            do_fail "The '#{step}' upgrade step failed: #{e.pretty_inspect}"
          end
        end
      else
        do_fail "The '#{step}' upgrade step is not implemented for this host"
      end
      return rc
    end

    # now the upgrade steps should look like:
    def run_upgrade_step_example(upgrade_state)
      # do some stuff...
      return 0 #or, return an error code if something went wrong
    end

    ######## UTILITIES ########

    def called_from; caller[1][/`([^']*)'/, 1]; end
    def eputs(msg); $stderr.write "\e[#{31}m#{msg}\e[0m\n"; end
    def wputs(msg); $stderr.write "\e[#{33}m#{msg}\e[0m\n"; end

    def debug(msg)
      return 0 unless @params[:debug]
      Logger.log(msg = "DEBUG: #{msg}\n")
      $stdout.write msg
      return 0
    end

    def verbose(msg)
      Logger.log(msg = "INFO: #{msg}\n")
      @params[:verbose] and $stdout.write msg
      return 0
    end

    def do_fail(msg)
      Logger.log(msg = "ERROR: #{called_from}\n" + msg)
      eputs msg
      return 1
    end

    def do_warn(msg)
      Logger.log(msg = "WARN: #{called_from}\n" + msg)
      wputs msg
      return 1
    end

    def run_script(script)
      script_status = @upgrade_state['run_scripts'] ||= {}
      if script_status[script]=='COMPLETE'
        verbose "skipping #{script} -- already ran successfully"
        return 0, ""
      end
      verbose "running #{script}"
      output = `#{script} 2>&1`
      rc = $?.exitstatus
      if $?.success?
        verbose "#{script} ran without error:\n--BEGIN OUTPUT--\n#{output}\n--END #{script} OUTPUT--"
        script_status[script] = 'COMPLETE'
      else
        do_fail "#{script} had errors:\n--BEGIN OUTPUT--\n#{output}\n--END #{script} OUTPUT--"
        script_status[script] = 'FAILED'
      end
      return rc, output
    end

    # shortcut for File.dirname(__FILE__) + script
    def run_script_relative(from, *script)
      run_script File.join(File.dirname(from), *script.map {|s| s.to_s})
    end

    # shortcut for File.dirname(__FILE__) + dir + scripts
    # If any fail (nonzero rc), execution ends
    def run_scripts_in(from, *dir)
      rc, output = 0, ""
      dir = File.join(File.dirname(from), *dir.map {|s| s.to_s})
      verbose "Running upgrade scripts in #{dir}"
      any_ran = false
      Dir.entries(dir).sort.each do |script|
        file = File.join dir, script
        next if File.directory? file
        next if !File.executable? file
        next if script.start_with? '.'
        any_ran = true
        rc, o = run_script(file)
        output += o
        break if rc != 0
      end
      if any_ran
        return rc, output
      else
        do_fail "no executable scripts found in #{dir} !!"
        return 1, output
      end
    end

  end

  class Logger
    LOG_FILE = ENV['OPENSHIFT_UPGRADE_LOG'] || '/var/log/openshift/upgrade.log'
    def self.file=(file)
      @file = file
    end
    def self.file
      @file ||= LOG_FILE
    end
    def self.log(msg)
      # not very efficient to open the file every time, but this is not
      # high-traffic so not a problem.
      File.open(self.file, 'a') {|f| f.write msg}
    end
  end #logger
end # OSEUpgrader module
